### DISCUSSION TITLE
.NET 9 Preview 6

### DISCUSSION THREAD

**Discussion by jamesmontemagno** [dotnet/core#9392] [News]
We are pleased to announce the release of .NET 9 Preview 6. 

This release contains the following improvements:

**📚Libraries:**
- [Improvements to System.Numerics](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#improvements-to-systemnumerics)
- [Support Primary Constructors in Logging Source Generator](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#support-primary-constructors-in-logging-source-generator)
- [System.Text.Json enhancements](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#systemtextjson) including a new [JsonSchemaExporter](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#jsonschemaexporter), [nullable annotations recognition](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#respecting-nullable-annotations), [requiring non-optional constructor parameters](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#requiring-non-optional-constructor-parameters), [ordering `JsonObject` properties](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#ordering-jsonobject-properties), and new [contract metadata APIs](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#additional-contract-metadata-apis)
- [`[GeneratedRegex]` can now be used on properties](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#generatedregex-on-properties)
- New [`EnumerateSplits`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#regexenumeratesplits) method for `Regex` to split more collection types
- Introduction of generic `OrderedDictionary` with [`OrderedDictionary<TKey, TValue>`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#ordereddictionarytkey-tvalue) 
- New [`ReadOnlySet<T>`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#readonlysett) when needing to create a read-only wrapper around `ISet<T>`
- [`allows ref struct` used in many places throughout the libraries](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#allows-ref-struct-used-in-many-places-throughout-the-libraries)
- [Collection lookups with spans](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#collection-lookups-with-spans)
- [More span-based APIs](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#more-span-based-apis) including `StartsWith` and `EndsWith` extension methods
- [Base64Url](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#base64url) for optimized encoding and decoding
- [SocketsHttpHandler by default in HttpClientFactory](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#socketshttphandler-by-default-in-httpclientfactory)
- [TLS resume with client certificates on Linux](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#tls-resume-with-client-certificates-on-linux)
- New [`System.Net.ServerSentEvents`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#systemnetserversentevents) library providing a parser for easily ingesting server-sent events
- [Introducing the Metrics Gauge Instrument](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md#introducing-the-metrics-gauge-instrument) in `System.Diagnostics.Metrics` to record non-additive values when changes occur
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/libraries.md)

**⏱️Runtime:**
- [ARM64 Code Generation](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/runtime.md#arm64-code-generation) now adds ability to store operations
- [Code Layout](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/runtime.md#code-layout) - RyuJIT's block reordering algorithm with a simpler, more global approach
- [Loop Optimizations](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/runtime.md#loop-optimizations) for code size reduction and performance improvements
- [Reduced Address Exposure](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/runtime.md#reduced-address-exposure) through RyuJIT improvements to better track usage of local variable address
- [AVX10v1 Support](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/runtime.md#avx10v1-support), a new SIMD instruction set from Intel
- [Hardware Intrinsic Code Generation](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/runtime.md#hardware-intrinsic-code-generation)
- [Constant Folding for Floating Point and SIMD Operations](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/runtime.md#constant-folding-for-floating-point-and-simd-operations)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/runtime.md)

**🛠️SDK**

- [NuGetAudit now raises warnings for vulnerabilities in transitive dependencies](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/sdk.md#nugetaudit-now-raises-warnings-for-vulnerabilities-in-transitive-dependencies)
- Addition of [`dotnet nuget why`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/sdk.md#dotnet-nuget-why) to find out why a transitive package is being used in your project
- [MSBuild BuildChecks](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/sdk.md#msbuild-buildchecks) to help users enforce rules and invariants during their builds
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/sdk.md)

You can find detailed release notes for additional features in .NET 9 Preview 6 below:

## C#

- [Partial properties](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/csharp.md#partial-properties)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/csharp.md)

## 🌐ASP.NET Core

- [Fingerprinting of static web assets](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md#fingerprinting-of-static-web-assets) ensuring that stale assets aren't used and enables improved caching behavior for faster load time
- [Improved distributed tracing for SignalR](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md#improved-distributed-tracing-for-signalr) with a new `ActivitySource`
- [Enhancements to Microsoft.AspNetCore.OpenAPI](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md#enhancements-to-microsoftaspnetcoreopenapi) including [completion enhancements](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md#completion-enhancements-and-package-install-recommendations-for-openapi-package), support for [`[Required]` and `[DefaultValue]` attributes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md#support-for-required-and-defaultvalue-attributes-on-parameters-or-properties), [schema transforms](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md#support-for-schema-transformers-on-openapi-document) on OpenAPI documents, 
- [Analyzer to warn when `[Authorize]` is overridden by `[AllowAnonymous]`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md#analyzer-to-warn-when-authorize-is-overridden-by-allowanonymous-from-farther-away), and new [analyzers](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md#analyzer-to-warn-when-authorize-is-overridden-by-allowanonymous-from-farther-away), 
- [`ComponentPlatform` renamed to `RendererInfo`](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md#componentplatform-renamed-to-rendererinfo)
- [Split large HTTP/2 headers across frames](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md#split-large-http2-headers-across-frames)
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/aspnetcore.md)

## .NET MAUI
- Full [release notes](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/dotnetmaui.md)

## Get started
To get started with .NET 9, [install the .NET 9 SDK](https://dotnet.microsoft.com/download/dotnet/9.0).

If you're on Windows using Visual Studio, we recommend installing the latest [Visual Studio 2022 preview](https://visualstudio.microsoft.com/vs/preview/), or get started with Visual Studio Code and the [C# Dev Kit](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csdevkit) extension.

You can find [.NET 9 Preview 6 Installer & Binaries](https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/9.0.0-preview.6.md) on GitHub.

## Team Announcements & Discussions
Discuss this release with the product teams on GitHub through the announcements for this release. 

* [.NET Libraries & Runtime](https://github.com/dotnet/runtime/discussions/104620)
* [.NET MAUI](https://github.com/dotnet/maui/discussions/23506) 
* [ASP.NET Core](https://github.com/dotnet/aspnetcore/discussions/56690)

## Stay up-to-date with .NET 9
You can stay up-to-date with all the features of .NET 9 with:

* [What's new in .NET 9](https://learn.microsoft.com/dotnet/core/whats-new/dotnet-9/overview)
* [What's new in C# 13](https://learn.microsoft.com/dotnet/csharp/whats-new/csharp-13)
* [What's new in ASP.NET Core](https://learn.microsoft.com/aspnet/core/release-notes/aspnetcore-9.0)
* [What's new in .NET MAUI](https://learn.microsoft.com/dotnet/maui/whats-new/dotnet-9)
* [What's new in EF Core](https://learn.microsoft.com/ef/core/what-is-new/ef-core-9.0/whatsnew)
* [Breaking Changes in .NET 9](https://learn.microsoft.com/dotnet/core/compatibility/9.0)
* [.NET 9 Releases](https://github.com/dotnet/core/blob/main/release-notes/9.0/README.md)

For .NET 9 previews we are utilizing GitHub Discussions to make announcements. For more information [read our discussion](https://github.com/dotnet/core/discussions/9131).

Stay up to date by subscribing to the [RSS feed](https://github.com/dotnet/core/discussions/categories/news.atom) for all release announcements.

**Comment by dannykay1710** [comment]
I really wish these updates were still on developer blogs, found it so much easier to keep upto date

**Reply by julealgon** [reply]
I disagree. It is a lot more convenient to subscribe to the github repo than it is to keep track of the blog.

Maybe they should have both, though.

**Reply by dannykay1710** [reply]
Yeah I get that. For me it was a single URL to goto for updates and memorable enough to access anywhere. 

I could look into how to subscribe to the preview discussions though, good shout.

**Reply by julealgon** [reply]
@dannykay1710 you can subscribe to specific types of posts too like this:
![image](https://github.com/user-attachments/assets/ccd11232-6456-4800-a1c1-2582b99d963b)
![image](https://github.com/user-attachments/assets/3301afbe-1670-4ba9-902d-47dbe205b71c)

**Reply by richlander** [reply]
Here's a single URL.

https://github.com/dotnet/core/discussions/9234

Also, we're going to look at changing the monthly servicing posts we do to just be an "updates" post and include a section about preview content.

https://devblogs.microsoft.com/dotnet/dotnet-and-dotnet-framework-july-2024-servicing-updates/

I want to raise something that is not at all obvious. With this new approach, there is a LOT more content being shared. The medium of blog posts demands content that is more curated and professional. We maintain a much lower bar for release notes. We still edit them and everything, however, we let the content go on much longer. Here's a perfect example: https://github.com/dotnet/core/blob/main/release-notes/9.0/preview/preview6/runtime.md#arm64-code-generation. There is a ton of great content there that the developer of the feature wrote. That content wasn't available before because it wasn't written. We (James, Jon, and myself) have been encouraging the team to "tell us everything". There are lots of other examples like that.

@jamesmontemagno @jongalloway

**Reply by dannykay1710** [reply]
Thanks for the feedback I really do appreciate it. Oddly though despite searching for the previews and updates via search engine I've never actually come across that index which is tremendously helpful.

I think upon reflection and the fact I agree there is a lot more content, my issue is less about its location and more about discoverability of the content "ad-hoc" which I'm not sure there is much you can do given it's fairly specific

**Reply by richlander** [reply]
Fully agree and get it.

Here's an even easier link that gets you to much the same information.

https://github.com/dotnet/core/blob/main/README.md#release-notes

That page gets you to this location: https://github.com/dotnet/core/blob/main/release-notes/9.0/README.md#release-notes

**Reply by richlander** [reply]
One more thing about release notes. We have had some people make PRs where they have found mistakes. Thanks! We want to encourage that behavior. In fact, I made such a PR just now: https://github.com/dotnet/core/pull/9406.

**Comment by colejohnson66** [comment]
The build check analyzer code links (BC0101 and BC0102) are 404s

**Reply by baronfel** [reply]
Fixed now - my PR https://github.com/dotnet/core/pull/9437 hadn't been merged.
